Issue/7847 reader site search

nbradbury opened the pull request:
Resolves #7797 - searching in the reader now includes site search results. Higher quality video demo [here](https://cloudup.com/cy8Ww_T5DF6).

![ezgif-2-b21ba344cc](https://user-images.githubusercontent.com/3903757/41163815-b874a792-6b07-11e8-99bf-bd65ae9a6c31.gif)

nozomimimi commented:
Wow, this is super! Looks great to me :) Thank you!

AmandaRiu commented on WordPress/src/main/java/org/wordpress/android/ui/reader/ReaderActivityLauncher.java:
```diff
@@ -86,15 +86,26 @@ public static void showReaderPostPagerForBlog(Context context,
-        if (siteId == 0) {
+        showReaderBlogPreview(context, siteId, 0);
+    }
+    public static void showReaderBlogPreview(Context context, long siteId, long feedId) {
```
Since this method now serves two functions, should the method name be updated to `showReaderBlogOrFeedPreview`? It's also not clear without reading the code that siteId should be 0 if you want to open the feed, and the old method `showReaderFeedPreview(context, feedId)` didn't require a siteId at all so this may cause some confusion. Doing something similar to the way you kept the original `showReaderBlogPreview(context, blogId)` and adding back in `showReaderFeedPreview(context, feedId)` and then just feeding that into the new combined method would go a long way to clear that up.

nbradbury replied:
In 08743c9 I renamed the method to `showReaderBlogOrFeedPreview` but I don't think a separate `showReaderFeedPreview` is necessary - at least, I'm not seeing a place where it's needed.

AmandaRiu commented on WordPress/src/main/java/org/wordpress/android/ui/reader/ReaderBlogFragment.java:
```diff
@@ -235,10 +235,6 @@ public void onBlogClicked(Object item) {
-        } else if (blogId != 0) {
-            ReaderActivityLauncher.showReaderBlogPreview(getActivity(), blogId);
-        }
+        ReaderActivityLauncher.showReaderBlogPreview(getActivity(), blogId, feedId);
```
This method can be simplified into:
```
@Override
public void onBlogClicked(Object item) {
    if (item instanceof ReaderRecommendedBlog) {
        ReaderRecommendedBlog blog = (ReaderRecommendedBlog) item;
        ReaderActivityLauncher.showReaderBlogPreview(getActivity(), blog.blogId);
    } else if (item instanceof ReaderBlog) {
        ReaderBlog blog = (ReaderBlog) item;
        ReaderActivityLauncher.showReaderBlogPreview(getActivity(), blog.blogId, blog.feedId);
    }
}
```

nbradbury replied:
Nice one! Also added in 08743c9.

AmandaRiu commented on WordPress/src/main/java/org/wordpress/android/ui/reader/ReaderPostListFragment.java:
```diff
@@ -643,8 +667,8 @@ public boolean onMenuItemActionExpand(MenuItem item) {
-                if (mBottonNavController != null) {
-                    mBottonNavController.onRequestHideBottomNavigation();
+                if (mBottomNavController != null) {
+                    mBottomNavController.onRequestHideBottomNavigation();
```
Took me a moment to see that change! Nice!

AmandaRiu commented:
@nbradbury 

The last item in the list is getting cut off for both tabs:
![screenshot-1528487763455](https://user-images.githubusercontent.com/5810477/41178239-ef943492-6b23-11e8-8abb-ec88eed0fe35.jpg)

Also, the tabs look a bit funny in landscape mode, I just want to verify this is by design:
![screenshot-1528487701881](https://user-images.githubusercontent.com/5810477/41178389-4bb32418-6b24-11e8-8369-069b384b6570.jpg)

Also as an android user I'm accustomed to being able to swipe between tabs but notice this is not available here. If not in this PR maybe we should open another ticket to make sure it gets added later?

AmandaRiu commented:
@nbradbury the ticket referenced in this PR is for "Move the reader filter to the app bar", is that the correct ticket?

AmandaRiu commented on WordPress/src/main/java/org/wordpress/android/ui/reader/ReaderPostListFragment.java:
```diff
@@ -725,6 +776,29 @@ private void submitSearchQuery(@NonNull String query) {
+
+        ReaderSiteSearchAdapter adapter = getSiteSearchAdapter();
+        if (event.isError()) {
+            adapter.clear();
```
Not sure how error handling is handled in this app...should the user see a notification?

nbradbury commented:
@AmandaRiu Thanks for the thorough review!

> The last item in the list is getting cut off for both tabs
> Rotate the device while and active search is happening, the progr ess indicator does not get enabled when the rotation is complete and the message says no results found even though the search is still happening
> Rotating the device after clearing the search results brings the search results back.

Yuck. All three of these are happening in `develop` as well (they're not specific to this PR). I filed issue #7882 to address them.

> Also, the tabs look a bit funny in landscape mode

I don't know why `TabLayout` does this because it's quite ugly. The only way to fix this was to use reflection, which I did in fbf2316.

> the ticket referenced in this PR is for "Move the reader filter to the app bar", is that the correct ticket?

Oops, corrected!

> Now that this method is renamed, we should move it below showReaderBlogPreview

Good call, addressed in 5792a3b.

> android:elevation doesn't work with AppBarLayout, it uses app:elevation. It looks like it's deprecated as of API 26

TIL! I ended up just removing the elevation from the XML in 7fb3ce0 and let Android apply the default elevation.

> Something really funky happened to the formatting of lines 112 - 174, the indentation has been removed. Same goes for lines 2070 - 2093. 

Weird, not sure how that happened. Fixed in 88ba8ac.

> the selected tab is not retained if activity needs to be recreated.

I'll work on this today and hopefully we can wrap things up afterwards.

nbradbury commented:
> Also as an android user I'm accustomed to being able to swipe between tabs but notice this is not available here. If not in this PR maybe we should open another ticket to make sure it gets added later?

Just realized I didn't respond to this, sorry! I felt that swiping (via a `ViewPager`) would've been harder to accomplish in the one week we were given to complete this task. I'll open a new issue for it, though, once this has been merged.

nbradbury commented:
@AmandaRiu Ready for another round.

AmandaRiu commented:
@nbradbury Regarding the changes in fbf2316cc08053d96df412dc5d094b325827a39b to fix the layout of the tabs in landscape mode, I've never been a fan of reflection due to the [performance hit](http://blog.nimbledroid.com/2016/02/23/slow-Android-reflection.html) so I did a little research and found an alternative if you're open to it. Instead of using reflection to modify `TabLayout_tabMaxWidth` which for some unknown reason is only configurable in XML, we could define the layout in XML and inflate it as needed. I pulled your old version using `git checkout 08743c9ac1823650568dd4e7825b12187c9bdcbb` and tested it out and the following changes worked:

Create a new xml layout file (I called mine `stretched_tab_layout.xml`)
```
<?xml version="1.0" encoding="utf-8"?>
<android.support.design.widget.TabLayout
    xmlns:android="http://schemas.android.com/apk/res/android"
    xmlns:app="http://schemas.android.com/apk/res-auto"
    android:layout_width="match_parent"
    android:layout_height="wrap_content"
    app:tabMaxWidth="0dp"
    app:tabGravity="fill"
    app:tabMode="fixed"/>

```

Update `ReaderPostListFragment` method (3 lines changed, 1 removed)
```
private void createSearchTabs() {
    if (mSearchTabs == null) {
        ViewGroup rootView = getView().findViewById(android.R.id.content);
        LayoutInflater inflater = LayoutInflater.from(getActivity());
        mSearchTabs = (TabLayout) inflater.inflate(R.layout.stretched_tab_layout, rootView);

        int indicatorColor = ContextCompat.getColor(getActivity(), R.color.tab_indicator);
        int textColor = ContextCompat.getColor(getActivity(), R.color.blue_light);
        int selectedTextColor = ContextCompat.getColor(getActivity(), R.color.white);

        mSearchTabs.setVisibility(View.GONE);
        mSearchTabs.setSelectedTabIndicatorColor(indicatorColor);
        mSearchTabs.setTabTextColors(textColor, selectedTextColor);

        mSearchTabs.addTab(mSearchTabs.newTab().setText(R.string.posts));
        mSearchTabs.addTab(mSearchTabs.newTab().setText(R.string.sites));

        mRecyclerView.getAppBarLayout().addView(mSearchTabs);
    }
}
```

Thoughts?

Here's a patch file if you want to try it out. Just make sure you checkout the commit I mentioned above: [Fix_tab_width.patch.zip](https://github.com/wordpress-mobile/WordPress-Android/files/2099168/Fix_tab_width.patch.zip)

AmandaRiu commented:
@nbradbury The fix to retain the selected tab in commit 653847ae9252bb60a9ca852568ef3ba0f1653393 works great, nicely done! However, the position in the list is not retained for the sites tab during orientation change and both tabs lose their scroll position if the activity needs to be recreated. I verified the main reader list does save this information so the lost state in search results would be a deviation from the users expected experience. Can we get this added? Since you're using a recyclerview retaining state is really straight forward and best done through the layout manager. I'm sure you probably already know this, but just in case here's a snippet I pulled from other work I did:
```
private var listState: Parcelable? = null // Save the state of the recycler view    

override fun onCreate(savedInstanceState: Bundle?) {
    super.onCreate(savedInstanceState)
    savedInstanceState?.let { bundle ->
        listState = bundle.getParcelable(STATE_KEY_LIST)
    }
}


override fun onActivityCreated(savedInstanceState: Bundle?) {
    super.onActivityCreated(savedInstanceState)

    listState?.let {
        ordersList.layoutManager.onRestoreInstanceState(listState)
        listState = null
    }
}    

override fun onSaveInstanceState(outState: Bundle) {
    val listState = ordersList.layoutManager.onSaveInstanceState()

    outState.putParcelable(STATE_KEY_LIST, listState)
    super.onSaveInstanceState(outState)
}
```

AmandaRiu commented:
@nbradbury Second round complete. Other than the last two comments, the rest looks great.

nbradbury commented:
> I've never been a fan of reflection due to the performance hit so I did a little research and found an alternative if you're open to it

I agree that's a better approach. In 7b351b4 I changed the search tabs to be inflated from XML.

I also fixed the saving & restoring of the site search position. I couldn't take the approach you suggested because the site search results are temporary (ie: after rotating a new search is performed). Instead I simply remember the position in `onSaveInstanceState()` and then restore once the search completes.

AmandaRiu commented:
@nbradbury Looks good! 👌 :shipit:
